### PR TITLE
Add Supervisor API to get field by index

### PR DESCRIPTION
**Pull request by stefaniapedrazzi**
Complete #3027 and synchronize with origin/develop branch:
* [x] minor fixes on the implementation
* [x] Add Matlab API
* [x] Add ROS API
* [x] Add documentation
* [x] Add tests

### CODE REVIEW

**Comment by stefaniapedrazzi** [comment]
Only ROS test failed but seems unrelated to the latest changes.

File: docs/reference/supervisor.md
Comment (lukicdarkoo) [inline docs/reference/supervisor.md:761]:
```suggestion
| `/supervisor/node/get_number_of_fields` | `service` | `webots_ros::node_get_number_of_fields` | `uint64 node`<br/>`bool proto`<br/>`---`<br/>`uint32 value` |
```

File: src/webots/nodes/utils/WbSupervisorUtilities.cpp
Comment (lukicdarkoo) [inline src/webots/nodes/utils/WbSupervisorUtilities.cpp:1170]: Should we call it ID or index?

File: src/webots/nodes/utils/WbSupervisorUtilities.cpp
Comment (lukicdarkoo) [inline src/webots/nodes/utils/WbSupervisorUtilities.cpp:298]: Not directly related to this PR, but this seems to be a bit cryptic, enum or similar would be helpful

**Review by lukicdarkoo — CHANGES_REQUESTED** [review]
Thank you, it looks good, there are only a few minor comments.

Don't forget to add an entry to the changelog

File: src/webots/nodes/utils/WbSupervisorUtilities.cpp
Comment (stefaniapedrazzi) [reply src/webots/nodes/utils/WbSupervisorUtilities.cpp:1170]: Index is the correct term for the public API: it represents the item at a given index in the node's list of field.
Fields objects don't have ids.
However the already existing code is (wrongly) referring to id also for fields (in `WbSupervisorUtilities.cpp` and `supervisor.c`).

In this PR I decided to not update/fix all the existing code variable and function names because it will cause many unrelated changes. But I can do it if you prefer.

File: src/webots/nodes/utils/WbSupervisorUtilities.cpp
Comment (stefaniapedrazzi) [reply src/webots/nodes/utils/WbSupervisorUtilities.cpp:298]: To me this is a normal behavior for variables meant to contain positive numbers like sizes or count:
`-1` means that it is not set, values `> 0` represent the found value, it is like NULL for pointers.
In any case, I don't see how you could use an enumeration to achieve the same result.
Note that any other solution would probably require additional variables.

Comment (lukicdarkoo) [reply src/webots/nodes/utils/WbSupervisorUtilities.cpp]: I was thinking something like this:
```cpp
enum { NOT_REQUESTED = -2, NOT_FOUND = -1 };

// ...

mFoundFieldId = NOT_REQUESTED;
```

Comment (stefaniapedrazzi) [reply src/webots/nodes/utils/WbSupervisorUtilities.cpp]: I'm not sure that mixing enum with int values is good.
With enumeration I would expect that all the possible values are listed in the enumeration and this won't be the case here.

If you want to add something like this, then I think you should not use an enumeration but simply define separate global variables for `NOT_REQUESTED` and `NOT_FOUND`.
However, this doesn't not only affect the two variables used in this PR but many others. So, if we want to improve this, it has to be done in a separate PR.

Comment (lukicdarkoo) [reply src/webots/nodes/utils/WbSupervisorUtilities.cpp]: > If you want to add something like this, then I think you should not use an enumeration but simply define separate global variables for `NOT_REQUESTED` and `NOT_FOUND`.

I agree, macros are probably a better idea.

> However, this doesn't not only affect the two variables used in this PR but many others. So, if we want to improve this, it has to be done in a separate PR.

👍

**Review by lukicdarkoo — APPROVED** [review]
Thank you

**Comment by stefaniapedrazzi** [comment]
I will wait for the the approval of https://github.com/cyberbotics/webots_ros/pull/71 before merging this PR.

**Pull request merged**
